osu.Game/Overlays/Chat/ChannelList/ChannelList.cs

```diff
@@ -0,0 +1,175 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Shapes;
+using osu.Framework.Input.Events;
+using osu.Game.Graphics;
+using osu.Game.Graphics.Containers;
+using osu.Game.Graphics.Sprites;
+using osu.Game.Online.Chat;
+using osuTK;
+
+namespace osu.Game.Overlays.Chat.ChannelList
+{
+    public class ChannelList : Container
+    {
+        public Action<Channel>? OnRequestSelect;
+        public Action<Channel>? OnRequestLeave;
+
+        public readonly BindableBool SelectorActive = new BindableBool();
+
+        private readonly Dictionary<Channel, ChannelListItem> channelMap = new Dictionary<Channel, ChannelListItem>();
+
+        private ChannelListItemFlow publicChannelFlow = null!;
+        private ChannelListItemFlow privateChannelFlow = null!;
+
+        [BackgroundDependencyLoader]
+        private void load(OverlayColourProvider colourProvider)
+        {
+            Children = new Drawable[]
+            {
+                new Box
+                {
+                    RelativeSizeAxes = Axes.Both,
+                    Colour = colourProvider.Background6,
+                },
+                new ChannelListScrollContainer
+                {
+                    Padding = new MarginPadding { Vertical = 7 },
+                    RelativeSizeAxes = Axes.Both,
+                    ScrollbarAnchor = Anchor.TopLeft,
+                    ScrollDistance = 35f,
+                    Child = new FillFlowContainer
+                    {
+                        Direction = FillDirection.Vertical,
+                        RelativeSizeAxes = Axes.X,
+                        AutoSizeAxes = Axes.Y,
+                        Children = new Drawable[]
+                        {
+                            publicChannelFlow = new ChannelListItemFlow("CHANNELS"),
+                            new ChannelListSelector
+                            {
+                                Margin = new MarginPadding { Bottom = 10 },
+                                SelectorActive = { BindTarget = SelectorActive },
+                            },
```

Comment:
With the way this selector is set up, you can have both an actual channel selected and the selector looking selected at the same time which is going to be confusing:

![2022-04-16-172938_364x732_scrot](https://user-images.githubusercontent.com/20418176/163680961-38e680c2-20df-4617-bf8b-2c41a5600f63.png)

This also [contradicts design](https://www.figma.com/file/f8b2dHp9LJCMOqYP4mdrPZ/Client%2FChat?node-id=2%3A63).

Reply:
I'm honestly not sure the best way to go about fixing this?
You could:
- Set the `Bindable<Channel> currentChannel` to `null` to indicate that no channel should be show
- Also pass the `BindableBool SelectorActive` to each of the items with some additional logic:
```csharp
            currentChannel.BindValueChanged(channelChange =>
            {
                if (channelChange.NewValue == channel && !SelectorActive.Value)
                    showSelect();
                else
                    hideSelect();
            }, true);

            SelectorActive.BindValueChanged(selectorChange =>
            {
                if (currentChannel.Value != channel)
                    return;

                if (selectorChange.NewValue)
                    hideSelect();
                else
                    showSelect();
            }, true);
```

I'm not particularly fond of either of these solutions :/

Reply:
Well I mean there isn't going to be a _nice_ way to write this with the current structure. Something like the above will be required.

Although I don't know why `currentChannel`'s value change callback is showing/hiding the selector, it should just be setting `SelectorActive`. Changes in visibility of the selector should occur only on changes to `SelectorActive`.

Reply:
Apologies, that snippet wasn't very clear.
`show/hideSelect` would be fading the `selectBox` within each `ChannelListItem`, not changing the state of the selector.

Reply:
I'm a bit confused. The intention on the design seems to be that when you are selecting channels, it *replaces* the currently displayed channel completely. ie. it *is* a channel "tab" itself and the previous selection is not remembered. Is that not how it's being implemented?

Reply:
You are correct in that it should work with tab like behaviour. This snippet is incorrect.
The existing `ChannelSelector` tabs in the current chat design use a `null` channel value to represent when the selector should be active.
I initially went with using a separate bindable bool to represent the selector state since I didn't feel that assigning null was good, but it seems it's actually simpler to do so in this case so I will switch to using that now.

Reply:
I've pushed e596c9d1713f82f3000f1c5029ee20d4a96ebdb3 to make the list behaviour work as expected.
This change introduces the enum `ChannelSelectorState` and a bindable that is intended to be cached at the chat overlay level.
Implementing it this way seemed the simplest way to get the desired behaviour while having a separate bindable for the current channel and the state of the selector.